internal/batch/slice.go

```diff
@@ -0,0 +1,143 @@
+package batch
+
+import (
+	"github.com/apache/arrow/go/v16/arrow"
+	"github.com/apache/arrow/go/v16/arrow/util"
+)
+
+type (
+	SlicedRecord struct {
+		arrow.Record
+		Bytes       int64 // we need this as the util.TotalRecordSize will report the full size even for the sliced record
+		bytesPerRow int64
+	}
+)
+
+func (s *SlicedRecord) split(limit *Cap) (add *SlicedRecord, toFlush []arrow.Record, rest *SlicedRecord) {
+	if s == nil {
+		return nil, nil, nil
+	}
+
+	add = s.getAdd(limit)
+	if add != nil {
+		limit.add(add.Bytes, add.NumRows())
+	}
+
+	if s.Record == nil {
+		// all processed
+		return add, nil, nil
+	}
+
+	toFlush = s.getToFlush(limit)
+	if s.Record == nil {
+		// all processed
+		return add, toFlush, nil
+	}
+
+	// set bytes & rows new values
+	limit.set(s.Bytes, s.NumRows())
+	return add, toFlush, s
+}
+
+func (s *SlicedRecord) getAdd(limit *Cap) *SlicedRecord {
+	rowsByBytes := limit.bytes.remainingPerN(s.bytesPerRow)
+	rows := limit.rows.remaining()
+	switch {
+	case rows < 0:
+		rows = rowsByBytes
+	case rows > rowsByBytes && rowsByBytes >= 0:
+		rows = rowsByBytes
+	}
+
+	switch {
+	case rows == 0:
+		return nil
+	case rows < 0, rows >= s.NumRows():
+		// grab the whole record (either no limits or not overflowing)
+		res := *s
+		s.Bytes = 0
+		s.Record = nil
+		return &res
+	}
+
+	res := SlicedRecord{
+		Record:      s.NewSlice(0, rows),
+		Bytes:       rows * s.bytesPerRow,
+		bytesPerRow: s.bytesPerRow,
+	}
+	s.Record = s.NewSlice(rows, s.NumRows())
+	s.Bytes -= res.Bytes
+	return &res
+}
+
+func (s *SlicedRecord) getToFlush(limit *Cap) []arrow.Record {
+	rowsByBytes := limit.bytes.capPerN(s.bytesPerRow)
+	rows := limit.rows.cap()
+	switch {
+	case rows < 0:
+		rows = rowsByBytes
+	case rows > rowsByBytes && rowsByBytes >= 0:
+		rows = rowsByBytes
+	}
+
+	switch {
+	case rows == 0:
+		// not even a single row fits
+		// we still need to process this, so slice by single row
+		return s.slice()
+	case rows < 0:
+		// as s.Record != nil we know that the limits are there in place & the s.Record.NumRows() > 0
+		panic("should never be here")
```

Comment:
If this can never happen what's the reason we have this case?

Reply:
we could remove it 💯 , just wanted to cover all possible cases (so if & when some code in-between changes we won't end up in mem corruption state)